docs/adr/0005-multitenancy-approach-for-marain.md

```diff
@@ -0,0 +1,263 @@
+# Multitenancy approach for Marain
+
+## Status
+
+Proposed
+
+## Context
+
+Tenancy has been made a first class citizen of all Marain services, however this by itself is not enough to make the system truly multitenanted. In order to do this, we need to determine how tenants should be created, managed and used within the Marain "world".
+
+We would like the option of deploying Marain as either a managed service, hosted by us and licenced to users as a PaaS offering, or for clients to deploy private instances into their own cloud subscriptions. We also want to give clients of the managed services the option for data to be stored in their own storage accounts or databases, but still have us run the compute aspects of the platform on their behalf.
+
+In addition, we need to be able to differentiate between a Marain service being available for a client to use directly and one being used as a dependency of a service they are using. For example, the Workflow service makes use of the Operations service. As a result, clients that are licenced to use the Workflow service will be using the Operations service indirectly, despite the fact that they may not be licenced to use it directly.
+
+We need to define a tenancy model that will support these scenarios and can be implemented using the `Marain.Tenancy` service.
+
+## Decision
+
+To support this, we have made the following decisions
+
+1. Every client using a Marain instance will have a top-level (i.e. child of the root tenant) Marain tenant created for them. For the remainder of this document, these will be referred to as "Client Tenants". Note that there is nothing that mandates these be top-level for this approach to work, but at present we expect that they will be.
+1. Every Marain service will also have a top-level Marain tenant created for it. For the remainder of this document, these will be referred to as "Service Tenants". As with Client Tenants, there is nothing that mandates they be top level.
```

Comment:
I thought we agreed these weren't "top-level" in that they're not directly under the root tenant. I thought there would be some sort of Marain Services tenant under the root, and each of these would be under that.

Reply:
Yes. We agreed that after I'd done this draft. Now updated to take account of this.